### DISCUSSION TITLE
Unable to resolve service for type OrchardCore.ContentManagement.Metadata.IContentDefinitionManager

### DISCUSSION THREAD

**Discussion by djampula4** [OrchardCMS/OrchardCore#16081] [Q&A]
I have created a class 'ProductPart' where these entities will be created in Sql DB.

```
 public class ProductPart : ContentPart
    {
        public decimal UnitPrice { get; set; }
        public string Sku { get; set; }
    }
```

To represent those entities in SQL we have to go through Migrations and the class look like the one below

```
 public class Migrations : DataMigration
    {
        private readonly IContentDefinitionManager _contentDefinitionManager;
        public Migrations(IContentDefinitionManager contentDefinitionManager)
        {
            this._contentDefinitionManager = contentDefinitionManager;
        }
        public int Create()
        {
            CreateTable();
            UpdateForm1();
            return 1;
        }

        private void CreateTable()
        {
            SchemaBuilder.CreateTable(nameof(ProductPart), table => table
           .Column<int>("Id", column => column.PrimaryKey().NotNull())
           .Column<decimal>("UnitPrice")
           .Column<string>("Sku", column => column.WithLength(50))
            );
        }

        private void UpdateForm1()
        {
            _contentDefinitionManager.AlterPartDefinition(nameof(ProductPart), part => 
                 part.Attachable()
                .WithDisplayName("Product")
                .WithDescription("Add products to the content."));
        }

        
    }
```
And in the startup class, I have created a scope instance, which is

```
 public void ConfigureServices(IServiceCollection services)
        {   
            services.AddContentPart<ProductPart>();
            services.AddScoped<IDataMigration, Migrations>();
            services.AddOrchardCms();
            
        }
```
Finally, when I ran this I am getting the below error.

*> Some services are not able to be constructed (Error while validating

the service descriptor 'ServiceType: OrchardCore.Data.Migration.IDataMigration Lifetime: Scoped ImplementationType: OrchardWebSiteModule.Migrations': Unable to resolve service for type 'OrchardCore.ContentManagement.Metadata.IContentDefinitionManager' while attempting to activate 'OrchardWebSiteModule.Migrations'.)*

**Comment by jtkech** [comment]
Normally the following need to be done from a **module** startup, so in the context of a tenant (we have at least the `Default` tenant)

            services.AddContentPart<ProductPart>();
            services.AddScoped<IDataMigration, Migrations>();

From the application that can behave as a module we have helpers e.g. to register services in a tenant container context

            services.AddOrchardCms()
                .ConfigureServices(tenantServices =>
                {
                     tenantServices.AddContentPart<ProductPart>();
                     ...
                });

**Comment by djampula4** [comment]
> Normally the following need to be done from a **module** startup, so in the context of a tenant (we have at least the `Default` tenant)
> 
> ```
>         services.AddContentPart<ProductPart>();
>         services.AddScoped<IDataMigration, Migrations>();
> ```
> 
> From the application that can behave as a module we have helpers e.g. to register services in a tenant container context
> 
> ```
>         services.AddOrchardCms()
>             .ConfigureServices(tenantServices =>
>             {
>                  tenantServices.AddContentPart<ProductPart>();
>                  ...
>             });
> ```

That even did not resolve my issue

**Comment by jtkech** [comment]
Hmm, i think it's a manifest dependency problem, see in our module manifest files how we define dependencies for a given feature, e.g.


    [assembly: Feature(
        Id = "OrchardCore.Contents.FileContentDefinition",
        Name = "File Content Definition",
        Description = "Stores Content Definition in a local file.",
        Dependencies = new[] { "OrchardCore.Contents" },
        Category = "Content Management"
    )]

You need to do the same with a dependency on `OrchardCore.ContentTypes`, hmm and as i remember the feature `ID` would need to be `YourWebAppAssemblyName.YourFeatureName`

And then, as i remember, your classes need to be decorated with the feature attribute

    [Feature("YourWebAppAssemblyName.YourFeatureName")]

But not sure of the details, only from memory

**Update**

Hmm, but there is a mix here, e.g. you don't need

        private void CreateTable()
        {
            SchemaBuilder.CreateTable(nameof(ProductPart), table => table
           .Column<int>("Id", column => column.PrimaryKey().NotNull())
           .Column<decimal>("UnitPrice")
           .Column<string>("Sku", column => column.WithLength(50))
            );
        }

As a part is intended to be used in a content type, and content items of this content type will be saved in the existing `Document` table

And the way to register part is different, e.g.

            // Autoroute Part
            services.AddContentPart<AutoroutePart>()
                .UseDisplayDriver<AutoroutePartDisplayDriver>() // <= Needs at least a driver
                .AddHandler<AutoroutePartHandler>(); // <= an handler is not required
